libbeat/licenser/manager.go

```diff
@@ -0,0 +1,279 @@
+// - Adding a watcher will automatically send the current license to the newly added watcher if
+//   available.
+type Manager struct {
+	done chan struct{}
+	sync.RWMutex
+	fetcher     Fetcher
+	duration    time.Duration
+	gracePeriod time.Duration
+	license     *License
+	watchers    map[Watcher]Watcher
+	log         *logp.Logger
+}
+
+// New takes an elasticsearch client and wraps it into a fetcher, the fetch will handle the JSON
+// and response code from the cluster.
+func New(client *elasticsearch.Client, duration time.Duration, gracePeriod time.Duration) *Manager {
+	fetcher := NewElasticFetcher(client)
+	return NewWithFetcher(fetcher, duration, gracePeriod)
+}
+
+// NewWithFetcher takes a fetcher and return a license manager.
+func NewWithFetcher(fetcher Fetcher, duration time.Duration, gracePeriod time.Duration) *Manager {
+	m := &Manager{
+		fetcher:     fetcher,
+		duration:    duration,
+		log:         logp.NewLogger("license-manager"),
+		done:        make(chan struct{}),
+		gracePeriod: gracePeriod,
+		watchers:    make(map[Watcher]Watcher),
+	}
+
+	return m
+}
+
+// AddWatcher register a new watcher to receive events when the license is retrieved or when the manager
+// is closed.
+func (m *Manager) AddWatcher(watcher Watcher) error {
+	m.Lock()
+	defer m.Unlock()
+
+	if _, ok := m.watchers[watcher]; ok {
+		return ErrWatcherAlreadyExist
+	}
+
+	m.watchers[watcher] = watcher
+
+	// when we register a new watchers send the current license unless we did not retrieve it.
+	if m.license != nil {
+		watcher.OnNewLicense(*m.license)
+	}
+	return nil
+}
+
+// RemoveWatcher removes the watcher if it exist or return an error.
+func (m *Manager) RemoveWatcher(watcher Watcher) error {
+	m.Lock()
+	defer m.Unlock()
+	if _, ok := m.watchers[watcher]; ok {
+		delete(m.watchers, watcher)
+		return nil
+	}
+	return ErrWatcherDoesntExist
+}
+
+// Get return the current active license, it can return an error if the manager is stopped or when
+// there is no license in the manager, Instead of querying the Manager it is easier to register a
+// watcher to listen to license change.
+func (m *Manager) Get() (*License, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	select {
+	case <-m.done:
+		return nil, ErrManagerStopped
+	default:
+		if m.license == nil {
+			return nil, ErrNoLicenseFound
+		}
+		return m.license, nil
+	}
+}
+
+// Start starts the License manager, the manager will start a go routine to periodically
+// retrieve the license from the fetcher.
+func (m *Manager) Start() {
+	// First update should be in sync at startup to ensure a
+	// consistent state.
+	m.log.Info("license manager started, no license found.")
+	go m.worker()
+}
+
+// Stop terminates the license manager, the go routine will be stopped and the cached license will
+// be removed and no more checks can be done on the manager.
+func (m *Manager) Stop() {
+	defer m.log.Info("license manager stopped")
```

Comment:
Should we have a lock on stop to make sure it's not called in the middle of a worker starting for example?

Reply:
This is a valid case, let me make a change.

Reply:
I refactored `Stop` so the shutdown is more explicit and sync with the worker, this makes the behavior of stopping the manager predictable and the callbacks are calling order is more sane.